### PR TITLE
[ENT-979] Use Enterprise Data API rather than Vertica for reporting.

### PR DESCRIPTION
**Pull request by UmanShahzad**
*Testing Instructions*:

1. As usual, set up all the proper Enterprise-related config in your LMS, e.g. an Enterprise Customer, a reporting config, etc.. Make sure the reporting config is for CSV `progress_v2` reports.
1. Clone this repo at this branch.
1. Set up a local version of https://github.com/edx/edx-analytics-data-api.
1.1. Clone the repo and make a virtual environment for it (python 2.7).
1.2. `make develop`
1.3. `pip install -Ue /path/to/edx-enterprise-data/on/your/machine` (the `make develop` command will have already installed a different version than this branch's, so we need to override it with this)
1.4. `./manage.py migrate --noinput`
1.5. `./manage.py migrate --noinput --database=analytics`
1.6. Temporarily change `analyticsdataserver.settings.local.LMS_BASE_URL` to have the proper port of your LMS, if necessary (e.g. change from 8000 to 18000 for the docker devstack).
1.7. Temporarily change `/path/to/edx-enterprise-data/enterprise_data/fixtures/enterprise_enrollment.json` to have the `enterprise_id` for the last 2 objects match your own Enterprise Customer's UUID.
1.8. `python manage.py loaddata /path/to/edx-enterprise-data/enterprise_data/fixtures/enterprise_enrollment.json --database analytics`
1.8. `./manage.py runserver` (if your LMS is on port 8000, you need to specify the port here).
2. As in the last PR, modify `enterprise_reporting.send_enterprise_reports.should_deliver_report` to `return True` by default.
3. Put `from pdb import set_trace;set_trace()` below `data_report_zipped = super().send(files)` inside of the delivery methods so you stop short of actually delivering anything.
4. Run `LMS_OAUTH_KEY=discovery-key LMS_OAUTH_SECRET=discovery-secret LMS_ROOT_URL=http://localhost:18000 LMS_OAUTH_HOST=http://localhost:18000 LMS_FERNET_KEY='DUMMY KEY CHANGE BEFORE GOING TO PRODUCTION' SEND_EMAIL_FROM='uman@mslm.me' ANALYTICS_API_URL=http://localhost:8000 python enterprise_reporting/send_enterprise_reports.py`
5. Check the `/tmp` directory for the zip file, open it up with the appropriate password, and ensure the columns and rows look good.

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @UmanShahzad! I've created [OSPR-2401](https://openedx.atlassian.net/browse/OSPR-2401) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-enterprise-data/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

**Comment by UmanShahzad** [comment]
Note that tests are failing because of a change to `caniusepython3` in the latest version (we don't pin). https://github.com/brettcannon/caniusepython3/pull/178 fixes it though, and the release should come soon.

**Comment by UmanShahzad** [comment]
@georgebabey This is ready for a code review now, tests finally passed. (Also probably ready to be tested on the Jenkins box; it worked for me locally (update: I forgot about the required URL though, nevermind)).

File: enterprise_reporting/reporter.py [hunk 1]
```diff
@@ -130,19 +101,13 @@ def _generate_enterprise_report(self):
...
-            data_report_file_writer.writerow(self.VERTICA_QUERY_FIELDS)
-            query = self.VERTICA_QUERY.format(
-                fields=','.join(self.VERTICA_QUERY_FIELDS),
-                enterprise_id=UUID(self.enterprise_customer_uuid).hex
-            )
-            LOGGER.debug('Executing this Vertica query: {}'.format(query))
-            data_report_file_writer.writerows(vertica_client.stream_results(query))
-        vertica_client.close_connection()
+            writer = csv.writer(data_report_file)
+            enrollments = EnterpriseDataApiClient().get_enterprise_enrollments(self.enterprise_customer_uuid)['results']
```
Comment (georgebabey) [inline enterprise_reporting/reporter.py]: nit: I'd probably move the enrollments `get` outside the `with open()` as to not keep the file handler longer than you need to.

**Review by georgebabey — APPROVED** [review]
Looks great, nice and straightforward.

File: enterprise_reporting/reporter.py (hunk 1 above)
Comment (pomegranited) [reply enterprise_reporting/reporter.py]: 👍  to this suggestion

**Review by pomegranited — APPROVED** [review]
:+1: After working out my JWT auth issues, this branch works as described.

However I did note that the order of the CSV headers for the two generated CSV files changes randomly every time the task is run.  E.g. on the first run, they were:
```
course_key,passed_timestamp,course_title,course_max_effort,course_pacing_type,lms_user_id,user_current_enrollment_mode,course_min_effort,user_username,course_duration_weeks,course_start,course_id,enterprise_id,enterprise_sso_uid,consent_granted,enterprise_name,enterprise_user_id,user_account_creation_timestamp,letter_grade,enrollment_created_timestamp,id,has_passed,user_email,course_end,enterprise_site_id
```
but on the second one they were:
```
enterprise_sso_uid,course_title,has_passed,id,user_email,course_max_effort,enterprise_name,passed_timestamp,course_end,course_id,course_min_effort,user_username,course_start,course_key,user_current_enrollment_mode,consent_granted,lms_user_id,enterprise_site_id,course_pacing_type,enterprise_id,enterprise_user_id,enrollment_created_timestamp,user_account_creation_timestamp,letter_grade,course_duration_weeks
```

This is unrelated to this change, but might be something that needs fixing if anyone is going to use these CSV reports?

- [x] I tested this on my devstack using the PR instructions.
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A
- [x] <strike>Includes documentation</strike> N/A

File: enterprise_reporting/reporter.py (hunk 1 above)
Comment (UmanShahzad) [reply enterprise_reporting/reporter.py]: :+1:

**Comment by UmanShahzad** [comment]
@pomegranited Thanks for the review! I'm surprised that is happening, thanks for noticing. I'll try to fix that before we move on.

**Comment by UmanShahzad** [comment]
@pomegranited OK, it should be fixed now. I've tested it to ensure it always comes out the same each time.

**Comment by UmanShahzad** [comment]
@georgebabey We just need the API url now and this is ready to be tested on the live Jenkins box.

**Comment by UmanShahzad** [comment]
@georgebabey This is ready to be tested.

File: enterprise_reporting/reporter.py [hunk 2]
```diff
@@ -130,19 +101,13 @@ def _generate_enterprise_report(self):
...
-            data_report_file_writer.writerow(self.VERTICA_QUERY_FIELDS)
-            query = self.VERTICA_QUERY.format(
-                fields=','.join(self.VERTICA_QUERY_FIELDS),
-                enterprise_id=UUID(self.enterprise_customer_uuid).hex
-            )
-            LOGGER.debug('Executing this Vertica query: {}'.format(query))
-            data_report_file_writer.writerows(vertica_client.stream_results(query))
-        vertica_client.close_connection()
+            writer = csv.writer(data_report_file)
+            writer.writerow(list(OrderedDict(sorted(enrollments[0].items())).keys()))
```
Comment (georgebabey) [inline enterprise_reporting/reporter.py]: This line is throwing an error when there were no enrollments returned `[]`

```
ERROR:__main__:Data report failed to send for Vandelay Industries (Direct Integration Test Company)
01:50:28 Traceback (most recent call last):
01:50:28   File "enterprise_reporting/send_enterprise_reports.py", line 35, in send_data
01:50:28     reporter.send_enterprise_report()
01:50:28   File "/edx/var/jenkins/shiningpanda/jobs/26c3a865/virtualenvs/26c3a865/lib/python3.5/site-packages/edx_enterprise_data-0.1.4-py3.5.egg/enterprise_reporting/reporter.py", line 85, in send_enterprise_report
01:50:28     files = self._generate_enterprise_report()
01:50:28   File "/edx/var/jenkins/shiningpanda/jobs/26c3a865/virtualenvs/26c3a865/lib/python3.5/site-packages/edx_enterprise_data-0.1.4-py3.5.egg/enterprise_reporting/reporter.py", line 100, in _generate_enterprise_report
01:50:28     ext=self.report_type
01:50:28   File "/edx/var/jenkins/shiningpanda/jobs/26c3a865/virtualenvs/26c3a865/lib/python3.5/site-packages/edx_enterprise_data-0.1.4-py3.5.egg/enterprise_reporting/reporter.py", line 108, in _generate_enterprise_report_progress_csv
01:50:28     writer.writerow(list(OrderedDict(sorted(enrollments[0].items())).keys()))
01:50:28 IndexError: list index out of range
```

We should guard clause this.

Comment (UmanShahzad) [reply enterprise_reporting/reporter.py]: :+1: sounds good.

Comment (UmanShahzad) [reply enterprise_reporting/reporter.py]: bcd71645507b66ebf05a5518b1991d138ed774fd should protect against this.

**Comment by UmanShahzad** [comment]
@georgebabey I've fixed the TravisCI issues on #36, so that should be good to merge. Let me know when, and I'll rebase on master, then we can test this.

**Comment by UmanShahzad** [comment]
@georgebabey Rebased on master now, should be ready to test.

**Comment by UmanShahzad** [comment]
@georgebabey Ready to go with the `progress_v2` flag from https://github.com/edx/edx-enterprise/pull/323.

**Comment by georgebabey** [comment]
Confirmed using jenkins pointing at this branch and testing both 'progress' and 'progress_v2' reports against our test company

**Comment by openedx-webhooks** [comment]
Although this pull request is already merged, I've created [OSPR-5541](https://openedx.atlassian.net/browse/OSPR-5541) so that we can track it in Jira.

There is nothing you have to do. No action is needed from your side.  Thanks again for your contribution.

**Pull request merged**